### PR TITLE
Add ability to modify datasets

### PR DESCRIPTION
**Pull request by stuartmcalpine**
## Add ability to modify datasets

In the schema yaml file there is now the option of a `modifiable` flag for each column, which by default is false, but, if true, users can update those columns with new values after the dataset has been registered.

For now only `dataset` and `execution` `description`'s can be updated.  

### Changes

- The generic `find_entry` function in the base table class can find entries from any table (CI test added for this function)
- The base table class now has a `modify` function which can alter any column that is "modifiable" (defined in schema yaml).
- Added CI test for modify (with this I've moved some of the generic test functions, e.g., `insert_dataset` into a common utils file that all tests can more easily use). The CI python test file is a bit long, I will refactor it in the next PR.
- Added `get_modifiable_columns` function for tables, e.g., `datareg.Registrar.dataset.get_modifiable_columns()`, which returns a list of all columns that can be modified. Added also CI test for function.
- Added a section for `modify` and `delete` functions to the first notebook tutorial. 

### Notes

- Need to think about what columns we want to allow users to modify.
- Need to think about typechecking, for now its only descriptions which are strings so thats ok.

### CODE REVIEW

File: src/dataregistry/registrar/base_table_class.py [hunk 1]
```diff
@@ -99,9 +106,39 @@ def modify(self, entry_id, modify_fields):
...
+            update_stmt = (
+                update(my_table)
+                .where(getattr(my_table.c, self.entry_id) == entry_id)
+                .values(modify_fields)
+            )
+            conn.execute(update_stmt)
+            conn.commit()
 
-    def find_entry(self, entry_id):
+    def find_entry(self, entry_id, only_valid=False, only_on_disk=False):
```
Comment (JoanneBogart) [inline src/dataregistry/registrar/base_table_class.py:141]: Not sure these are the right defaults; something to discuss.

**Comment by JoanneBogart** [comment]
Concerning which fields should be modifiable - I like your approach of adding a "modifiable" field to the column description although I can't at the moment come up with any columns other than "description".  However there are several columns I can imagine users wanting to modify because they made a mistake the first time around. Here a concept of "force" could be useful.  There are probably some columns which should not be modified (like maybe "name" or "relpath") under any circumstances. If someone got it wrong they should delete the entry and start over.

File: src/dataregistry/registrar/base_table_class.py (hunk 1 above)
Comment (stuartmcalpine) [reply src/dataregistry/registrar/base_table_class.py]: Yes, this was largely me being lazy to reduce the checks at the other end.

The spirit of this function should just be to find an entry regardless of other factors.

I've removed the `only_valid` and `only_on_disk` options. I've added a `raise_if_not_found=False` optional tag to raise an exception if the dataset is not found.

I've moved the "is it already deleted" check back to the `delete` function.

**Comment by stuartmcalpine** [comment]
> Concerning which fields should be modifiable - I like your approach of adding a "modifiable" field to the column description although I can't at the moment come up with any columns other than "description". However there are several columns I can imagine users wanting to modify because they made a mistake the first time around. Here a concept of "force" could be useful. There are probably some columns which should not be modified (like maybe "name" or "relpath") under any circumstances. If someone got it wrong they should delete the entry and start over.

Potentially these columns could be modifiable:

### Execution table
- description
- execution_start
- locale
- configuration

### Dataset
- the version and version suffix (as long as it doesn't clash)
- creation_date
- access_API
- description
- is_overwritable

There is also `owner` and `owner_type` that could be changed. But these are potentially more similar to the promotion between schemas, like changing the `owner` would be like `transfer`.

**Review by JoanneBogart — APPROVED** [review]
This is fine for now.   In the near-future we should decide

- whether two values (True, False) for the `modifiable` field in the schema are sufficient or whether there should be a third, which allows a field to be modified only if something like `force` is specified
- exactly which fields should be in each category

**Pull request merged**